### DISCUSSION TITLE
Trying to add support for arctis 7+ headset

### DISCUSSION THREAD

**Discussion by MrNavaStar** [Sapd/HeadsetControl#203] [Q&A]
Ive created a new device and it is properly detected. (most of my code is based off of the other arctis headsets) The first feature I am trying to implement is sidetone.

Wire shark returns the following data when setting sidetone:
``` 39 01 00 00 00 ...... wlen = 64```
The second value has a range of 0 to 3

I tried the following code, but it doesn't work:
```c
static int arctis_7_plus_send_sidetone(hid_device* device_handle, uint8_t num)
{
    int ret = -1;

    // the range of the Arctis 7+ seems to be from 0 to 0x03 (3)
    num = map(num, 0, 3, 0x00, 0x03);

    unsigned char* buf = calloc(64, 1);

    if (!buf) {
        return ret;
    }

    const unsigned char data_on[2]  = { 0x39, num };
    const unsigned char data_off[1] = { 0x39 };

    if (num) {
        memmove(buf, data_on, sizeof(data_on));
    } else {
        memmove(buf, data_off, sizeof(data_off));
    }

    ret = hid_write(device_handle, buf, 64);

    free(buf);

    if (ret >= 0) {
        ret = arctis_7_plus_save_state(device_handle);
    }

    return ret;
}
```
Any thoughts?

**Comment by Sapd** [comment]
> Wire shark returns the following data when setting sidetone:
> ` 39 01 00 00 00 ...... wlen = 64`
> The second value has a range of 0 to 3

What is the exact data you get? (Also all the fields including all headers like wValue etc.)

**Reply by MrNavaStar** [reply]
I get the following data when setting side tone:

- bmRequestType: 0x21
- bRequest: SET_REPORT (0x09)
- wValue: 0x0200
- wIndex: 3
- wLength: 64

And the data fragment is 39, followed by the value, and then 62 zeros

**Reply by Sapd** [reply]
It is a SET_REPORT request, so you must use the `hid_send_feature_report` function, instead of hid_write

**Reply by MrNavaStar** [reply]
Okay I tried swapping ```hid_write``` with ```hid_send_feature_report``` but now I get ```Error: -1: ioctl (SFEATURE): Broken pipe```

**Reply by Sapd** [reply]
hmm wValue ends with a 00, so you probably must use 0x0 as first byte, e.g.

```
const unsigned char data_on[3]  = { 0x0, 0x39, num }
```

**Reply by Sapd** [reply]
Also I just saw, wIndex is 3 (interfaceID), you may have to set it in the init function:
```
thedevice.capability_details[CAP_SIDETONE]           = (struct capability_detail) {  .interface = 3 };
```

**Reply by MrNavaStar** [reply]
Hey no errors! but setting the sidetone doesn't seem to be working... Any idea what might cause something like that?

Or does map take an input of 1-128 and map it to 0-4?

**Reply by Sapd** [reply]
You can try the different values.
What happens if you again leave out the leading 0 in data_on? (so with size 2) with the correct interfaceid now?

**Reply by MrNavaStar** [reply]
It broke when I tried removing the leading 0

**Reply by Sapd** [reply]
Okay just looked into it, so the 0 in front is correct. Which platform are you? Windows or Mac/Linux?

On Mac/Linux it should definitely work like you did.

**Reply by MrNavaStar** [reply]
I'm on Linux. I get the success message when I run it, but side tone isn't working

**Reply by Sapd** [reply]
Hm maybe also watch it with WireShark and compare whats different.

You can also experiment with the dev menu:
```
./headsetcontrol --dev -- --device VENDORID:PRODUCTID --interface 3 --send-report "0x00 0x39 0x01"
```

**Reply by MrNavaStar** [reply]
the dev menu cant seem to find --send-report

**Reply by Sapd** [reply]
Ah there is a typo in the readme. It's actually '--send-feature' 

It won't da anything different than in code, but it will help you testing without recompiling every time

**Reply by MrNavaStar** [reply]
Sorry I haven't responded in a while, I'm still stuck tho lol. --send-feature was right, but i was never able to get sidetone to turn on even tho I was getting a success message. Im not sure what else to do. I tried comparing with wireshark again but no luck. Would i maybe be able to send the wireshark log for you to look at?

**Reply by MrNavaStar** [reply]
okay wait I found something odd. I managed to set up wireshark on linux and found that the wValue is 0x0300 instead of 0x0200. also the wLength is 63 instead of 64, even tho they are set to 64 in the code

**Reply by MrNavaStar** [reply]
YOOOOOO I got it!!! turns out i was supposed to use hid_write not hid_send_feature_report! ngl idk why

**Comment by MicaelJarniac** [comment]
What's the progress on this? I'm looking into getting a new headset, and the Arctis 7+ is looking promising.

**Reply by MrNavaStar** [reply]
Oh I got side tone working, I'll have to spend some more time to get the other stuff done. I can probably try and do it either this or next week.

**Reply by MicaelJarniac** [reply]
Any idea if it'd be compatible with the 7P+? I can't find the 7+ to buy anywhere, but the 7P+ is quite easy to find.

**Reply by MrNavaStar** [reply]
no clue, but I can give you a hand with wire shark if your willing to get your hands dirty.

On another note, the latest release of wine says it fixed support for steelseries gg. I'm not sure if it actually works but you could install the software and see.

**Reply by MicaelJarniac** [reply]
I ended up buying another headset, the Astro A50.

I was between that one, the Arctis 7+ and the Arctis Pro Wireless.

The 7+ I couldn't find anywhere (thus, I was looking into the 7P+). The Pro Wireless I managed to find a few, but way too expensive, and was almost impossible to find new ones. The A50 was more readily available, and according to reviews, it was supposed to be very good.

**Reply by CalebQ42** [reply]
I'd definitely be interested in it, I just got a 7+ and would love to be able to configure it on Linux.

**Reply by CalebQ42** [reply]
I'm also more then willing to help, if you need anything
